sdk/communication/AzureCommunicationCalling/ApiSamples.md

```diff
@@ -0,0 +1,758 @@
+    if (error == nil) {
+        print("Local video stopped successfully")
+    } else {
+        print("Local video failed to stop")
+    }
+}
+```
+
+### Local Video Preview
+You can use Renderer to start render stream from your local camera, this stream won't be send to other participants, it's local preview feed
+* [Synchronous] Start local video preview
+```swift
+let firstCamera: VideoDeviceInfo? = self.deviceManager?.cameras?.first
+let localVideoStream = LocalVideoStream(camera: firstCamera)
+let previewRenderer = try! Renderer(localVideoStream: localVideoStream)
+let previewView = try! previewRenderer!.createView(with: RenderingOptions(scalingMode:scalingMode))
+
+rendererObserver = RendererObserver(view: self)
+previewView!.delegate = rendererObserver
+```
+
+```swift
+public class RendererObserver: NSObject, RendererDelegate {
+    let view: View
+    init(view: View) {
+        self.view = view
+    }
+
+    public func onFirstFrameRendered() {
+        print("Size of frame is " + view.renderer!.size)
+    }
+
+    public func rendererFailedToStart() {
+        print("Rederer failed to start")
+    }
+}
+```
+
+Renderer has set of properties and methods that allows you to control it:
+```swift
+// Constructor can take in LocalVideoStream or RemoteVideoStream
+let localRenderer = Renderer(localVideoStream:localVideoStream)
+let remoteRenderer = Renderer(remoteVideoStream:remoteVideoStream)
+
+// [StreamSize] size of the rendering view
+renderer?.size
+
+// [RendererDelegate] an object you provide to receive events from this Renderer instance
+renderer?.delegate
+
+// [Synchronous] create view with rendering options
+renderer?.createView(with: RenderingOptions(scalingMode:ScalingMode.crop));
+// [Synchronous] dispose rendering view
+renderer?.dispose();
+```
+
+### Switch Video Source
+* [Asynchronous] Switch local video, pass localVideoStream you got from call.startVideo() API call
+```swift
+// get camera
+var firstCamera = self.deviceManager?.cameras?.first
+
+localVideoStream?.switchSource(camera: firstCamera) { (error) in
+    if (error == nil) {
+        print("Successfully switched to new source")
+    } else {
+        print("Failed to switch to new source")
+    }
+})
+```
+<!--End Video Basics-->
+
+## Device Management
+<!--Start Device Management-->
+### Microphone / Speaker
+Device mananager allows you to set device that will be used when starting a call
+If not  will fallback to OS defaults
+
+#### Get / Set Devices
+```swift
+// get microphone
+var firstMicrophone = self.deviceManager?.microphones?.first
+
+// get speaker
+var firstSpeaker = self.deviceManager?.speakers?.first
+
+// [Synchronous] set microphone
+self.deviceManager?.setMicrophone(microphoneDevice: firstMicrophone)
+
+// [Synchronous] set speaker
+self.deviceManager?.setSpeaker(speakerDevice: firstSpeaker)
+```
+
+AudioDeviceInfo has a set of properties:
+```swift
+var audioDeviceInfo = self.deviceManager?.microphones?.first
```

Comment:
Same comment here does this make sense for microphones to be nullable? I would expect that if the device manager didn't have any to just return an empty array, or throw

Reply:
I wonder if this is the result of ObjC->Swift bridging, but there should be annotations to control this

Reply:
We still don't have support for nullability type specifiers. ApiGen at the moment does not allow us to specify type specifier for property types, this was already mentioned that we are working on this. Its a big implementation in ApiGen so we havent got around  that yet, but will be fixed before GA.